runtime/macros/src/lib.rs

```diff
@@ -1,140 +1,11 @@
-// based on riscv-rt
-
-#![deny(warnings)]
-
-extern crate proc_macro;
-#[macro_use]
-extern crate quote;
-extern crate core;
-extern crate proc_macro2;
-#[macro_use]
-extern crate syn;
-
-use proc_macro2::Span;
-use syn::{
-    parse, punctuated::Punctuated, spanned::Spanned, FnArg, ItemFn, Lit, Meta, NestedMeta,
-    PathArguments, ReturnType, Type, Visibility,
-};
-
 use proc_macro::TokenStream;
 
+mod entry;
+mod parse_args;
+
 #[proc_macro_attribute]
 pub fn main(args: TokenStream, input: TokenStream) -> TokenStream {
-    let f = parse_macro_input!(input as ItemFn);
-    let a = parse_macro_input!(args with Punctuated::<Meta, syn::Token![,]>::parse_terminated);
-
-    // check the function arguments
-    if f.sig.inputs.len() > 3 {
-        return parse::Error::new(
-            f.sig.inputs.last().unwrap().span(),
-            "function has too many arguments",
-        )
-        .to_compile_error()
-        .into();
-    }
-    for arg in &f.sig.inputs {
-        match arg {
-            FnArg::Receiver(_) => {
-                return parse::Error::new(arg.span(), "invalid argument")
-                    .to_compile_error()
-                    .into();
-            }
-            FnArg::Typed(t) => {
-                if !is_simple_type(&t.ty, "u32") {
-                    return parse::Error::new(t.ty.span(), "argument type must be usize")
-                        .to_compile_error()
-                        .into();
-                }
-            }
-        }
-    }
-
-    // check the function signature
-    let valid_signature = f.sig.constness.is_none()
-        && f.sig.asyncness.is_none()
-        && f.vis == Visibility::Inherited
-        && f.sig.abi.is_none()
-        && f.sig.generics.params.is_empty()
-        && f.sig.generics.where_clause.is_none()
-        && f.sig.variadic.is_none()
-        && match f.sig.output {
-            ReturnType::Default => true,
-            ReturnType::Type(_, ref ty) => is_simple_type(ty, "u32"),
-        };
```

Comment:
This is excessive. For example, changing `main` visibility is fine, generics would cause an error on their own, empty `where` clause is harmless etc.

On native, users will face an error from original requirements from `fn main()`. On riscv target most of these do nothing.

Only sensible thing to do is to disable arguments and return value.

Reply:
Isn't that kinda of the idea of this code though? To affirmatively throw and give the user a clean error message rather than have some code/building silently behave differently than on native?

I guess more generally I'm unclear on why "excessive" is a problem here. It's essentially costless and enforces a clear contract of what code the VM is intended to run. Seems fine to me to leave this in.

Reply:
Rust gives you a nice error when you change signature of `main`, but this code doesn't, it's a blind copy-paste.
(note that original descriptive message is propagated by default, unless doing `nexus run`)

My original comments explains the reasoning. If we wan, for example, prohibit usage of generics which would never compile anyway, we should give an explicit error 
```
`main` function is not allowed to have generic parameters
```

etc. We can do it, but I don't see why. Maybe prohibiting `async` is important.

Reply:
A note:

An easy way to check expected signature (similar to how Rust does this) is to encode the following into the macro:
```rust
let func = parse_macro_input!(input as syn::ItemFn);
let fn_name = &func.sig.ident;
quote! { 
    #func

    const fn check_main(_: fn()) {}
    const _: () = {
        // Gives a compile error if main is unsafe, has ABI, or any inputs/outputs
        check_main(#fn_name)
    };
}.into()
```

But this isn't desirable until we can optimize this const away with release build

Reply:
So maybe I misunderstood your original comment, but I took what you wrote to mean that on `riscv` targets you don't necessarily get the nice error message, but instead most of these changes just invisibly stop working.

If you get the nice error message either way then just ignore this comment entirely. I'd just rather explicitly throw a nice error message we write than have different behavior depending on arch.

Reply:
https://github.com/nexus-xyz/nexus-zkvm/pull/195#discussion_r1648860164 the above comment explains how to match macro behavior. At the cost of more steps in debug mode:(